Sanitizer error in ydb/tests/functional/dstool/test_canonical_requests.py.Test.test_pdisk_set_status_inactive

ydb/tests/functional/dstool/test_canonical_requests.py.Test.test_pdisk_set_status_inactive

**Build Information:**
- Version: 1d75130d8825c8312517727a05f40306b2375fe7
- Build Preset: release-tsan
- Branch: prestable-26-1

**Test Report:** [View report](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/Regression-run_Small_and_Medium/21834037548/ya-prestable-26-1-x86-64-tsan/try_1/ya-test.html#FAIL)

**Run Command:**
```
./ya make -ttt --build "release" --sanitize="thread" -DDEBUGINFO_LINES_ONLY -F 'test_canonical_requests.py::Test::test_pdisk_set_status_inactive' ydb/tests/functional/dstool
```

**Owner:** [TEAM:@ydb-platform/blobstorage](https://github.com/orgs/ydb-platform/teams/blobstorage)

**Test run history:** [link](https://datalens.yandex/34xnbsom67hcq?branch=prestable-26-1&full_name=ydb/tests/functional/dstool/test_canonical_requests.py.Test.test_pdisk_set_status_inactive&build_type=release-tsan)

More info in [dashboard](https://datalens.yandex/4un3zdm0zcnyr)

**Logs:**
- [log](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/Regression-run_Small_and_Medium/21834037548/ya-prestable-26-1-x86-64-tsan/try_1/artifacts/logs/ydb/tests/functional/dstool/test-results/py3test/testing_out_stuff/test_canonical_requests.py.Test.test_pdisk_set_status_inactive.log)
- [logsdir](https://storage.yandexcloud.net/ydb-gh-logs/ydb-platform/ydb/Regression-run_Small_and_Medium/21834037548/ya-prestable-26-1-x86-64-tsan/try_1/test_artifacts/ydb-tests-functional-dstool.zip)

**Error Summary:**
```
E   warning: address range table at offset 0x880 has a premature terminator entry at offset 0x890
E   ==================
E   WARNING: ThreadSanitizer: data race (pid=904178)
E   Read of size 8 at 0x7254000924c8 by thread T269 (mutexes: write M0):
E   #0 NKikimr::TStreamCypher::SetKey(unsigned long const&) /-S/ydb/core/blobstorage/crypto/crypto.cpp:209:14 (ydbd+0x2c681fc0) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #1 SetKey /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_crypto.h:86:18 (ydbd+0x2c5903db) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #2 NKikimr::NPDisk::TSectorWriter<true, true>::TSectorWriter(NKikimr::TPDiskMon&, NKikimr::NPDisk::IBlockDevice&, NKikimr::NPDisk::TDiskFormat&, unsigned long&, unsigned long const&, NKikimr::NPDisk::TBufferPool*, unsigned long, unsigned long, unsigned long, unsigned int, NKikimr::NPDisk::TLogChunkInfo*, unsigned long, NKikimr::NPDisk::TBuffer*, std::__y1::shared_ptr<NKikimr::NPDisk::TPDiskCtx>, NKikimr::NPDisk::TDriveModel*, bool) /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_writer.h:139:16 (ydbd+0x2c5903db)
E   #3 NKikimr::NPDisk::TPDisk::WriteApplyFormatRecord(NKikimr::NPDisk::TDiskFormat, unsigned long const&) /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_impl.cpp:1763:23 (ydbd+0x2c58fad5) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #4 NKikimr::NPDisk::TPDisk::WriteDiskFormat(unsigned long, unsigned int, unsigned int, unsigned long const&, unsigned long const&, unsigned long const&, unsigned long const&, unsigned long const&, TBasicString<char, std::__y1::char_traits<char>>, bool, bool, std::__y1::optional<TRcBuf>, bool, std::__y1::optional<bool>) /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_impl.cpp:1852:5 (ydbd+0x2c591ea4) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #5 NKikimr::FormatPDisk(TBasicString<char, std::__y1::char_traits<char>>, unsigned long, unsigned int, unsigned int, unsigned long const&, unsigned long const&, unsigned long const&, unsigned long const&, unsigned long const&, TBasicString<char, std::__y1::char_traits<char>>, NKikimr::TFormatOptions const&) /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_tools.cpp:121:12 (ydbd+0x2c542d71) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #6 NKikimr::NPDisk::TPDiskActor::InitHandle(TAutoPtr<NActors::TEventHandle<NKikimr::NPDisk::TEvPDiskMetadataLoaded>, TDelete>)::'lambda'(void*)::operator()(void*) const /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp:462:29 (ydbd+0x2c8946b5) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #7 NKikimr::NPDisk::TPDiskActor::InitHandle(TAutoPtr<NActors::TEventHandle<NKikimr::NPDisk::TEvPDiskMetadataLoaded>, TDelete>)::'lambda'(void*)::__invoke(void*) /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp:426:17 (ydbd+0x2c89359d) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #8 (anonymous namespace)::TPosixThread::ThreadProxy(void*) /-S/util/system/thread.cpp:245:20 (ydbd+0x19fd5770) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   
E   Previous write of size 8 at 0x7254000924c8 by main thread:
E   #0 __tsan_memset /-S/contrib/libs/clang20-rt/lib/tsan/rtl/tsan_interceptors_memintrinsics.cpp:35:3 (ydbd+0x19d70acd) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #1 SecureWipeBuffer /-S/ydb/core/blobstorage/crypto/secured_block.h:40:5 (ydbd+0x2c804c1c) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #2 NKikimr::NPDisk::TPDiskActor::~TPDiskActor() /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp:213:9 (ydbd+0x2c804c1c)
E   #3 NKikimr::NPDisk::TPDiskActor::~TPDiskActor() /-S/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp:212:20 (ydbd+0x2c8053a9) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #4 CleanupActor /-S/ydb/library/actors/core/mailbox_lockfree.cpp:271:9 (ydbd+0x1b60042a) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #5 NActors::TMailbox::CleanupActors() /-S/ydb/library/actors/core/mailbox_lockfree.cpp:288:21 (ydbd+0x1b60042a)
E   #6 Cleanup /-S/ydb/library/actors/core/mailbox_lockfree.cpp:355:27 (ydbd+0x1b602e7d) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #7 NActors::TMailboxTable::Cleanup() /-S/ydb/library/actors/core/mailbox_lockfree.cpp:677:48 (ydbd+0x1b602e7d)
E   #8 NActors::TExecutorPoolBaseMailboxed::Cleanup() /-S/ydb/library/actors/core/executor_pool_base.cpp:281:30 (ydbd+0x1b5ab4b3) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #9 NActors::TCpuManager::Shutdown() /-S/ydb/library/actors/core/cpu_manager.cpp:170:40 (ydbd+0x1b56a73c) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #10 NActors::TActorSystem::Stop() /-S/ydb/library/actors/core/actorsystem.cpp:471:21 (ydbd+0x1b541715) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #11 NKikimr::TKikimrRunner::KikimrStop(bool) /-S/ydb/core/driver_lib/run/run.cpp:2281:22 (ydbd+0x2ddcc40e) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #12 NKikimr::MainRun(NKikimr::TKikimrRunConfig const&, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:52:17 (ydbd+0x2c285e9c) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #13 NKikimr::NDriverClient::TClientCommandServer::Run(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/core/driver_lib/cli_utils/cli_cmds_server.cpp:51:12 (ydbd+0x2dcacdbe) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #14 NYdb::NConsoleClient::TClientCommand::ValidateAndRun(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:290:16 (ydbd+0x2c3ab53a) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #15 NYdb::NConsoleClient::TClientCommandTree::Run(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:478:33 (ydbd+0x2c3ade8b) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #16 NYdb::NConsoleClient::TClientCommand::ValidateAndRun(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:290:16 (ydbd+0x2c3ab53a) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #17 NYdb::NConsoleClient::TClientCommand::Process(NYdb::NConsoleClient::TClientCommand::TConfig&) /-S/ydb/public/lib/ydb_cli/common/command.cpp:232:16 (ydbd+0x2c3aa70d) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #18 NKikimr::NDriverClient::NewClient(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/cli_utils/cli_cmds_root.cpp:77:26 (ydbd+0x2c3a2687) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #19 NKikimr::Main(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:153:20 (ydbd+0x2c28a22d) (BuildId: db45829858c759758e2185a5a049269ac571f4fb)
E   #20 ParameterizedMain(int, char**, std::__y1::shared_ptr<NKikimr::TModuleFactories>) /-S/ydb/core/driver_lib/run/main.cpp:203:16 (ydbd+0x2c28ba03) (BuildId: db45829858c759758e2185a5a049269...```

--- a/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp
+++ b/ydb/core/blobstorage/pdisk/blobstorage_pdisk_actor.cpp
@@ -210,6 +210,9 @@ public:
     }
 
     ~TPDiskActor() {
+        if (FormattingThread) {
+            FormattingThread->Join();
+        }
         SecureWipeBuffer((ui8*)MainKey.Keys.data(), sizeof(NPDisk::TKey) * MainKey.Keys.size());
     }
 
@@ -421,14 +424,12 @@ public:
         P_LOG(PRI_WARN, BSP01, PDisk->ErrorStr);
 
         // Is used to pass parameters into formatting thread, because TThread can pass only void*
-        using TCookieType = std::tuple<TPDiskActor*, TActorSystem*, TActorId, std::optional<TRcBuf>>;
+        using TCookieType = std::tuple<TIntrusivePtr<TPDiskConfig>, NPDisk::TKey, TActorSystem*, TActorId, std::optional<TRcBuf>>;
         FormattingThread.Reset(new TThread(
                 [] (void *cookie) -> void* {
                     auto params = static_cast<TCookieType*>(cookie);
-                    auto [actor, actorSystem, pDiskActor, metadata] = *params;
+                    auto [cfg, mainKey, actorSystem, pDiskActor, metadata] = std::move(*params);
                     delete params;
-
-                    TPDiskConfig *cfg = actor->Cfg.Get();
 
                     if (cfg->ReadOnly) {
                         TString readOnlyError = "PDisk is in read-only mode";
@@ -456,11 +457,11 @@ public:
 
                         try {
                             FormatPDisk(cfg->GetDevicePath(), 0, cfg->SectorSize, cfg->ChunkSize,
-                                cfg->PDiskGuid, chunkKey, logKey, sysLogKey, actor->MainKey.Keys.back(), TString(),
+                                cfg->PDiskGuid, chunkKey, logKey, sysLogKey, mainKey, TString(),
                                 options);
                         } catch (NPDisk::TPDiskFormatBigChunkException) {
                             FormatPDisk(cfg->GetDevicePath(), 0, cfg->SectorSize, NPDisk::SmallDiskMaximumChunkSize,
-                                cfg->PDiskGuid, chunkKey, logKey, sysLogKey, actor->MainKey.Keys.back(), TString(),
+                                cfg->PDiskGuid, chunkKey, logKey, sysLogKey, mainKey, TString(),
                                 options);
                         }
                         actorSystem->Send(pDiskActor, new TEvPDiskFormattingFinished(true, ""));
@@ -470,7 +471,7 @@ public:
                     }
                     return nullptr;
                 },
-                new TCookieType(this, TlsActivationContext->ActorSystem(), SelfId(), std::move(ev->Get()->Metadata))));
+                new TCookieType(Cfg, MainKey.Keys.back(), TlsActivationContext->ActorSystem(), SelfId(), std::move(ev->Get()->Metadata))));
 
         FormattingThread->Start();
     }

